server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java

```diff
@@ -148,47 +159,74 @@ void cancelTaskAndDescendants(CancellableTask task, String reason, boolean waitF
+        final TaskId parentTaskId = new TaskId(clusterService.localNode().getId(), task.getId());
+        sendBanParentRequests(childConnections, ActionListener.wrap(() -> {}),
+            subNodes -> new BanParentTaskRequest(parentTaskId, subNodes));
+    }
+
+    private void sendBanParentRequests(Map<String, List<DiscoveryNode>> childConnections, ActionListener<Void> listener,
+                                       Function<List<DiscoveryNode>, BanParentTaskRequest> requestGenerator) {
+        if (childConnections.isEmpty()) {
             listener.onResponse(null);
             return;
         }
-        logger.trace("cancelling task {} on child nodes {}", task.getId(), childNodes);
-        GroupedActionListener<Void> groupedListener =
-            new GroupedActionListener<>(ActionListener.map(listener, r -> null), childNodes.size());
-        final BanParentTaskRequest banRequest = BanParentTaskRequest.createSetBanParentTaskRequest(
-            new TaskId(clusterService.localNode().getId(), task.getId()), reason, waitForCompletion);
-        for (DiscoveryNode node : childNodes) {
-            transportService.sendRequest(node, BAN_PARENT_ACTION_NAME, banRequest,
-                new EmptyTransportResponseHandler(ThreadPool.Names.SAME) {
-                    @Override
-                    public void handleResponse(TransportResponse.Empty response) {
-                        groupedListener.onResponse(null);
-                    }
-
-                    @Override
-                    public void handleException(TransportException exp) {
-                        assert ExceptionsHelper.unwrapCause(exp) instanceof ElasticsearchSecurityException == false;
-                        logger.warn("Cannot send ban for tasks with the parent [{}] to the node [{}]", banRequest.parentTaskId, node);
-                        groupedListener.onFailure(exp);
-                    }
-                });
+        final int groupSize = childConnections.entrySet().stream()
+            .mapToInt(e -> e.getKey().equals(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY) ? e.getValue().size() : 1)
+            .sum();
+        final GroupedActionListener<Void> groupedListener =
+            new GroupedActionListener<>(ActionListener.map(listener, r -> null), groupSize);
+        for (Map.Entry<String, List<DiscoveryNode>> entry : childConnections.entrySet()) {
+            final String clusterAlias = entry.getKey();
+            if (clusterAlias.equals(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY)) {
+                final BanParentTaskRequest request = requestGenerator.apply(List.of());
+                for (DiscoveryNode node : entry.getValue()) {
+                    sendBanParentRequest(() -> transportService.getConnection(node), request, groupedListener);
+                }
+            } else {
+                final ArrayList<DiscoveryNode> subNodes = new ArrayList<>(entry.getValue());
+                final DiscoveryNode targetNode = subNodes.remove(0);
```

Comment:
I wonder if batching by cluster here is overly heroic (as it makes the whole down-stream logic remote-cluster aware)?

We will only be sending these ban requests for any pending original requests that are passing cluster boundary, so it might not be too bad?

For the heartbeat component, we can still do some kind of batching?

The advantage of the connection-less approach here, however, is that it allows a cluster to disconnect, subsequently reconnect, and possibly continue to send heartbeats.

Reply:
I agreed. I've simplified it in https://github.com/elastic/elasticsearch/pull/55779/commits/85081f51dc130b6522ae698a94263602b7e7cbac.